### PR TITLE
Azure flow logs from storage accounts

### PR DESCRIPTION
**Pull request by shireesh-illumio**
Add module to onboard azure flow storage accounts, and example on how to use it.

Fix links in readme to submodules in terraform module page (now showing 404).

Add outputs to the `azure_subscription` module with its variable values.

### CODE REVIEW

File: modules/azure_flow_logs_storage_accounts/variables.tf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  type = string
+  description = "The client ID of the service principal used to access the storage accounts."
+  validation {
+    condition     = length(var.service_principal_client_id) > 0
+    error_message = "The service_principal_client_id value must not be empty."
+  }
+}
+
+variable "storage_accounts" {
+  type    = list(object({
```
Comment (rlenglet) [inline modules/azure_flow_logs_storage_accounts/variables.tf:21]: Can't this be a `set`?

File: examples/azure_flow_logs_storage_accounts/versions.tf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+terraform {
+  required_providers {
+    illumio-cloudsecure = {
+      source  = "illumio/illumio-cloudsecure"
+      version = ">= 1.1.0"
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/versions.tf:5]: The `azure_flow_logs_storage_account` resource got added in 1.2.0.
```suggestion
      version = ">= 1.2.0"
```

File: examples/azure_flow_logs_storage_accounts/main.tf [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  ]
+}
+
+module "azure_flow_logs_storage_account_dev" {
+  source = "illumio/cloudsecure/illumio//modules/azure_flow_logs_storage_account"
+  version = "1.4.0"
+  service_principal_client_id = module.azure_subscription_dev.service_principal_id
+  storage_accounts = [
+    {
+      name               = "bucket_name"
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/main.tf:33]:
```suggestion
      name                = "bucket_name"
```

File: examples/azure_flow_logs_storage_accounts/main.tf [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      resource_group_name = "bucket_resource_group_name"
+    }
+  ]
+}
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/main.tf:37]: Add `tags`.

File: examples/azure_flow_logs_storage_accounts/main.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  tenant_id              = "b563cc25-a007-4837-981a-cbe2017228a4" # Azure Tenant ID
+
+  tags = [
+    "Environment=Dev",
+    "Owner=John Doe"
+  ]
+}
+
+module "azure_flow_logs_storage_account_dev" {
+  source = "illumio/cloudsecure/illumio//modules/azure_flow_logs_storage_account"
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/main.tf:28]:
```suggestion
  source = "illumio/cloudsecure/illumio//modules/azure_flow_logs_storage_accounts"
```

File: examples/azure_flow_logs_storage_accounts/main.tf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  subscription_id        = "1681e851-ba2d-410b-a66a-9511887e1c1a" # Azure Subscription ID
+  tenant_id              = "b563cc25-a007-4837-981a-cbe2017228a4" # Azure Tenant ID
+
+  tags = [
+    "Environment=Dev",
+    "Owner=John Doe"
+  ]
+}
+
+module "azure_flow_logs_storage_account_dev" {
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/main.tf:27]:
```suggestion
module "azure_flow_logs_storage_accounts" {
```

**Review by rlenglet — CHANGES_REQUESTED** [review]
Looks good overall.
I just have a few relatively minor comments.

File: modules/azure_flow_logs_storage_accounts/variables.tf
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_flow_logs_storage_accounts/variables.tf:1]: none of the resource in the module can accommodate tags. 
the one you pointed out was a data resource.

File: modules/azure_flow_logs_storage_accounts/variables.tf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    name = string
+    resource_group_name = string
+  }))
+  description = "The list of storage accounts to be monitored."
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_flow_logs_storage_accounts/variables.tf]: we are using data source to fetch the accounts which might be having flow logs.

File: modules/azure_subscription/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,19 @@
+output "azure_secret_expiration_days" {
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_subscription/outputs.tf]: thank you

File: modules/azure_subscription/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  value = var.iam_name_prefix
+  description = "The prefix given to all Azure resource names."
+}
+
+output "mode" {
+  value = var.mode
+  description = "The account's access mode, must be \"ReadWrite\" (default) or \"Read\"."
+}
+
+output "service_principal_id" {
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_subscription/outputs.tf]: thank you

File: modules/azure_subscription/outputs.tf
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_subscription/outputs.tf:1]: users when using the module can use these default applied values to test their code with whats created on aws. Agree it needs to be at all places. will do so in another pr.

File: modules/azure_flow_logs_storage_accounts/main.tf [hunk 10]
```diff
@@ -0,0 +1,52 @@
+data "azurerm_subscription" "current" {}
+
+locals {
+  storage_accounts_map = {
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_flow_logs_storage_accounts/main.tf]: changed to set in variable itself. not needed anymore.

File: modules/azure_flow_logs_storage_accounts/main.tf [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    for sa in var.storage_accounts : "${sa.name}-${sa.resource_group_name}" => sa
+  }
+}
+
+# Retrieve storage account scopes (for existing storage accounts)
+data "azurerm_storage_account" "account" {
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_flow_logs_storage_accounts/main.tf:3]: this is a data resource.

File: modules/azure_flow_logs_storage_accounts/main.tf [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  for_each          = local.storage_accounts_map
+  name                = each.value.name
```
[the remark this answers is not in the document]
Comment (shireesh-illumio) [reply modules/azure_flow_logs_storage_accounts/main.tf:5]: data resource.

File: modules/azure_subscription/outputs.tf [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  description = "The prefix given to all Azure resource names."
+}
+
+output "mode" {
+  value       = var.mode
+  description = "The account's access mode, must be \"ReadWrite\" (default) or \"Read\"."
+}
+
+output "secret_expiration_days" {
+  value       = var.azure_secret_expiration_days
```
Comment (rlenglet) [inline modules/azure_subscription/outputs.tf:17]: At least use the same name for consistency.
```suggestion
output "azure_secret_expiration_days" {
  value       = var.azure_secret_expiration_days
```

File: modules/azure_subscription/outputs.tf [hunk 14]
```diff
@@ -0,0 +1,19 @@
+output "service_principal_id" {
+  value       = azuread_service_principal.illumio_sp.object_id
+  description = "The ID of the service principal created for Illumio CloudSecure."
+}
```
Comment (rlenglet) [inline modules/azure_subscription/outputs.tf:4]: This is meant to be used as input for `azure_flow_logs_storage_accounts`'s `service_principal_client_id` variable. Use the same name to make that obvious.
```suggestion
output "service_principal_client_id" {
  value       = azuread_service_principal.illumio_sp.object_id
  description = "The ID of the service principal created for Illumio CloudSecure."
}
```

File: examples/azure_flow_logs_storage_accounts/main.tf [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  tags = [
+    "Environment=Dev",
+    "Owner=John Doe"
+  ]
+}
+
+module "azure_flow_logs_storage_accounts" {
+  source                      = "illumio/cloudsecure/illumio//modules/azure_flow_logs_storage_accounts"
+  version                     = "1.4.0"
+  service_principal_client_id = module.azure_subscription_dev.service_principal_id
```
Comment (rlenglet) [inline examples/azure_flow_logs_storage_accounts/main.tf:30]: Fix here after renaming.
```suggestion
  service_principal_client_id = module.azure_subscription_dev.service_principal_client_id
```

File: modules/azure_flow_logs_storage_accounts/variables.tf [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  default     = "IllumioCloudIntegration"
+  validation {
+    condition     = length(var.iam_name_prefix) > 0
+    error_message = "The iam_name_prefix value must not be empty."
+  }
+}
+
+variable "service_principal_client_id" {
+  type        = string
+  description = "The client ID of the service principal used to access the storage accounts."
```
Comment (rlenglet) [inline modules/azure_flow_logs_storage_accounts/variables.tf:13]: Clarify a bit.
```suggestion
  description = "The client ID of the service principal used by CloudSecure to access the storage accounts."
```

File: modules/azure_flow_logs_storage_accounts/variables.tf [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    error_message = "The service_principal_client_id value must not be empty."
+  }
+}
+
+variable "storage_accounts" {
+  type = set(object({
+    name                = string
+    resource_group_name = string
+  }))
+  description = "The set of storage accounts to retrieve information for."
```
Comment (rlenglet) [inline modules/azure_flow_logs_storage_accounts/variables.tf:25]: Reword to be consistent with `aws_flow_logs_s3_buckets`:
/variables.tf
```suggestion
  description = "The set of storage accounts containing flow logs."
```

File: modules/azure_flow_logs_storage_accounts/main.tf [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    ]
+    data_actions = [
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read"
+    ]
+  }
+
+  assignable_scopes = ["/subscriptions/${data.azurerm_subscription.current.subscription_id}"]
+  scope             = "/subscriptions/${data.azurerm_subscription.current.subscription_id}"
+}
+
```
Comment (rlenglet) [inline modules/azure_flow_logs_storage_accounts/main.tf:27]: Remove this extra empty line.
```suggestion
```

**Review by rlenglet — CHANGES_REQUESTED** [review]
Looks great! I only have a few minor comments.

**Review by rlenglet — DISMISSED** [review]
LGTM!!

**Review by rlenglet — APPROVED** [review]

**Review by rlenglet — APPROVED** [review]

**Pull request merged**